oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/MetadataRegistry.java

```diff
@@ -173,24 +172,31 @@ public Measure registerMeasureModel(Model model, BanyanDBStorageConfig config, C
             schemaBuilder.field(field.getName());
         }
         // parse TopN
-        schemaBuilder.topNSpec(parseTopNSpec(model, schemaMetadata.name()));
+        schemaBuilder.topNSpec(parseTopNSpec(model, schemaMetadata.name(), tagsAndFields));
 
         registry.put(schemaMetadata.name(), schemaBuilder.build());
         return builder.build();
     }
 
-    private TopNSpec parseTopNSpec(final Model model, final String measureName) {
-        final String valueCName = ValueColumnMetadata.INSTANCE.getValueCName(model.getName());
-        if (StringUtil.isEmpty(valueCName)) {
+    private TopNSpec parseTopNSpec(final Model model, final String measureName, MeasureMetadata tagsAndFields) {
+        final Optional<ValueColumnMetadata.ValueColumn> valueColumnOpt = ValueColumnMetadata.INSTANCE.readValueColumnDefinition(model.getName());
+        if (valueColumnOpt.isEmpty()) {
             return null;
         }
-        // TODO: how to configure parameters?
+
+        List<String> groupByTagNames = new ArrayList<>();
+        groupByTagNames.add(Metrics.ENTITY_ID);
+        for (final TagMetadata tagSpec : tagsAndFields.tags) {
+            if (tagSpec.getTagSpec().getTagName().equals(InstanceTraffic.SERVICE_ID)) {
+                groupByTagNames.add(InstanceTraffic.SERVICE_ID);
```

Comment:
How about adding an `Array` in the `TopNAggregation` annotation to define the columns? Then we could verify whether these columns are valid. Only should boot successfully after validation passed.

Reply:
If we add an array to `TopNAggregation`, do we need to annotate all Metrics that should have TopN support?

Reply:
All single value metrics, I think. We don't support labeled value metrics for topN query.

That is why the manual setup is better than this hidden logic.

BTW, the top-N query support includes two cases, w/ or w/o parent service ID. Usually instance Top-N should have service ID but it is not required from protocol level. Also, service metric has no parent for topN

Reply:
> All single value metrics, I think. We don't support labeled value metrics for topN query.
> 
> That is why the manual setup is better than this hidden logic.
> 
> BTW, the top-N query support includes two cases, w/ or w/o parent service ID. Usually instance Top-N should have service ID but it is not required from protocol level. Also, service metric has no parent for topN

I've added `@TopNAggregation` for

- OAL: dynamically annotate `@TopNAggregation` for single-valued metrics,
- MAL: only annotate `@MeterFunction`s that generate single value